pkg/ha/ha_service.go

```diff
@@ -108,5 +110,78 @@ func (ha *HAService) Start(ctx context.Context) error {
 		}
 	}
 
-	return nil
+	<-ctx.Done()
+	err := ha.cleanUpServiceEndpoints()
+	if err == nil {
+		err = ctx.Err()
+	}
+	return err
+}
+
+// cleanUpServiceEndpoints is executed upon ending leadership. Its purpose is to remove the Endpoints object created upon acquiring
+// leadership.
+func (ha *HAService) cleanUpServiceEndpoints() error {
+	// Use our own context. This function executes when the main application context is closed.
+	// Also, try to finish before a potential 15 seconds termination grace timeout.
+	ctx, cancel := context.WithTimeout(context.Background(), 14*time.Second)
+	defer cancel()
+	seedClient := ha.manager.GetClient()
+
+	attempt := 0
+	var err error
+	for {
+		endpoints := corev1.Endpoints{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      app.Name,
+				Namespace: ha.namespace,
+			},
+		}
+		err = seedClient.Get(ctx, client.ObjectKeyFromObject(&endpoints), &endpoints)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				ha.log.V(app.VerbosityVerbose).Info("The endpoints object cleanup succeeded: the object was missing")
+				return nil
+			}
+
+			ha.log.V(app.VerbosityInfo).Info("Failed to retrieve the endpoints object", "error", err.Error())
+		} else {
+			// Avoid data race. We don't want to delete the endpoint if it is sending traffic to a replica other than this one.
+			if !ha.isEndpointStillPointingToOurReplica(&endpoints) {
+				// Someone else is using the endpoint. We can't perform safe cleanup. Abandon the object.
+				ha.log.V(app.VerbosityWarning).Info(
+					"Abandoning endpoints object because it was modified by an external actor")
+				return nil
+			}
+
+			// Only delete the endpoint if it is the resource version for which we confirmed that it points to us.
+			deletionPrecondition := client.Preconditions{UID: &endpoints.UID, ResourceVersion: &endpoints.ResourceVersion}
+			err = seedClient.Delete(ctx, &endpoints, deletionPrecondition)
+			if client.IgnoreNotFound(err) == nil {
+				// The endpoint was deleted (even if not by us). We call that successful cleanup.
+				ha.log.V(app.VerbosityVerbose).Info("The endpoints object cleanup succeeded")
+				return nil
+			}
+			ha.log.V(app.VerbosityInfo).Info("Failed to delete the endpoints object", "error", err.Error())
+		}
+
+		// Deletion request failed, possibly because of a midair collision. Wait a bit and retry.
+		attempt++
+		if attempt >= 10 {
+			break
+		}
+		time.Sleep(1 * time.Second)
```

Comment:
What about using a [time.NewTicker](https://pkg.go.dev/time#NewTicker), or better yet [`clock.RealClock.NewTicker`](https://pkg.go.dev/k8s.io/utils/clock#RealClock.NewTicker) which allows you to use a `Clock` interface that can be mocked for tests.
Tickers take into account the time that was actually spent while executing the `Get`/`Update` calls. Additionally, since you will have to add a `select` statement for it, you can use the `select` to check if the context has expired meanwhile